File: Dynamic Programming/kadane/python/kadane.py
def kadane(a,n):
    ans=max(a)
    s=0
    for e in a:
        s+=e
        if s<0:
            s=0
        ans = max(ans,s)

    return ans


n=int(input('Enter the size of array'+'\n'))
a=[int(i) for i in input().split()]
print(kadane(a,n))

File: Dynamic Programming/kadane/kadane.cpp
#include <iostream>
#include <climits>
using namespace std;
 
#define MAX(X, Y) (X > Y) ? X : Y
#define POS(X) (X > 0) ? X : 0
 
int maxSum = INT_MIN;
int N;
int kadane(int* row, int len)
{
    int x, sum, maxSum = INT_MIN;
    for (sum = POS(row[0]), x = 0; x < N; ++x, sum = POS(sum + row[x]))
        maxSum = MAX(sum, maxSum);
    return maxSum;
}
 
int main()
{
    cout << "Enter the array length: ";
    cin >> N;
    int arr[N];
    cout << "Enter the array: ";
    for (int i = 0; i < N; i++)
    {
        cin >> arr[i];
    }
    cout << "The Max Sum is: "<<kadane(arr, N) << endl;
    return 0;
}
